### ISSUE TITLE
Higher capacity in Mega Man models

### ISSUE THREAD

**Issue by schrum2** [schrum2/MarioDiffusion#200]
I just made this issue for Mario Maker:
https://github.com/schrum2/MarioMakerPCG/issues/25

I essentially want you to experiment with all the same things in Mega Man. The important things to be aware of:

- Make batch files, don't change train_diffusion.py
- Must have 32 by 32 input samples if we are going to have 4 dim_mult layers

**Comment by schrum2** [comment]
Once the 32x32 dataset is made, create a batch file that is something like this one:
https://github.com/schrum2/MarioMakerPCG/blob/main/bat/train-mario-maker-big-unconditional.bat

Then train an unconditional model on that data.

I will also train a model with some capacity improvements on the older MMLV data with 16x14 samples

**Comment by fwilson12** [comment]
I'm currently pulling levels from MM Maker to generate enough samples to train with, but everything is in place to start training right away once the levels are downloaded

**Comment by schrum2** [comment]
I'm currently running this command on my office machine:
```
python train_diffusion.py --batch_size 8 --model_dim 256 --dim_mults 1 2 4 4 --attention_head_dim 8 --down_block_types "DownBlock2D" "AttnDownBlock2D" "AttnDownBlock2D" "AttnDownBlock2D" --up_block_types "AttnUpBlock2D" "AttnUpBlock2D" "AttnUpBlock2D" "UpBlock2D" --game MM-Full --save_image_epochs 20 --output_dir "BIG_CAPACITY-MM-unconditional-MMLVdata" --num_epochs 500 --json datasets\MM_LevelsAndCaptions-filtered-regular.json --seed 0 --mixed_precision bf16
```

It turns out the increasing the model capacity really ups the memory footprint, so I had to limit things compared to what was proposed in the Mario Maker issue. Specifically:
- `--mixed_precision bf16` included (instead of 32bit floating point)
- `--dim_mults 1 2 4 4`: Probably the biggest change. Having 8 as the last value introduced too many extra parameters.
- `--batch_size 8`: This is really small, and I worry it could hurt. Consider revisitin this later.
- `--attention_head_dim 8`: way smaller than originally desired, but at least there are still Attn blocks that take some advantage of this now.

We'll see how well this trains. @fwilson12 will need to make some similar cuts before training on the 32x32 dataset. However, I can probably allow a bigger model when I train on my home machine later.

**Comment by fwilson12** [comment]
I'm seeing the same spontaneous spike in loss that happened during the concurrent Mario Maker run:

##

<img width="1000" height="600" alt="Image" src="https://github.com/user-attachments/assets/b3dbbe8c-b39f-4caa-9a57-fd4f0cb145c3" />

### epoch 10 samples 

<img width="909" height="269" alt="Image" src="https://github.com/user-attachments/assets/5d798e03-379c-4cfd-9e86-8206daec6c61" />

### epoch 20 samples

<img width="910" height="271" alt="Image" src="https://github.com/user-attachments/assets/a6aa7818-0697-4e49-8b59-546de1f09756" />

### epoch 30 samples :(

 
<img width="1072" height="271" alt="Image" src="https://github.com/user-attachments/assets/ee0b7b0c-ee39-49fe-958c-93278bd6de90" />

**Comment by schrum2** [comment]
Both here and in https://github.com/schrum2/MarioMakerPCG/issues/25 we think that gradient clipping might solve the issue. Runs in progress now.

**Comment by fwilson12** [comment]
The unconditional model I was training is done; it doesn't look terrible and it didn't run into the loss-spike issue, but the levels are largely incoherent and filled with spikes:
##

<img width="1098" height="279" alt="Image" src="https://github.com/user-attachments/assets/87149fab-afa6-4d19-981d-8c8037c5029d" />

##

Will make a slightly larger dataset and try training again

**Comment by schrum2** [comment]
Adding myself to this issue as a reminder that I should try training an even larger capacity model on my home machine

**Comment by fwilson12** [comment]
Just pulled another ~1k levels from MM Maker, just started training a new model with 10k training samples.

**Comment by schrum2** [comment]
Please put the dataset in Google Drive and provide link

**Comment by fwilson12** [comment]
https://drive.google.com/drive/folders/1O4VLBP-YclXoDvoNbTcz-P6AMuB5b-an

**Comment by schrum2** [comment]
The dataset above is a large dataset (~10k samples) of size 32x32 that we are currently training an unconditional model for with a higher capacity UNet.

Something that we should also try: create a large dataset (~20k) samples of size 16x16. We can train both unconditional models and text-conditioned models on this. The text conditioned models can rely on our deterministic captions (which Alaa has been improving, so merge her branch) and also on LLM captions. The reason this issue is in this thread is that I want to try training a higher capacity text-conditioned model, though the risk of exceeding memory is even higher in this case. I might have to train the model at home.

In any case, don't create any new datasets from MMLV until issue #199 is resolved.

**Comment by fwilson12** [comment]
the second round of training an unconditional model on the 32x32 MMLV scenes wasn't much better despite increasing the size of the training set:

##

<img width="1103" height="312" alt="Image" src="https://github.com/user-attachments/assets/247e968a-04ed-4427-a1b0-4af9e2437239" />

##

Next step is to try training a model with deterministically-captioned 16x16 MMLV levels (#202)

**Comment by schrum2** [comment]
Issue #202 will involve training a text-conditioned model with the default UNet capacity, which is why it is a distinct issue from this one. I'm going to leave this issue open for now. If training for Issue #202 already puts us near the max VRAM threshold, then we'll just close this big capacity issue and give up on it, but if there is still remaining VRAM, we can try training a high capacity conditional model (possibly on my home machine).

**Comment by schrum2** [comment]
I just killed the run in my office. This was the one mentioned above ( https://github.com/schrum2/MarioDiffusion/issues/200#issuecomment-4836141300 ) that was using standard single room scenes and an unconditional higher capacity model. The training is super slow, and doesn't seem to be producing anything special (epoch 240):

<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/dcbd96f7-72ac-4081-8d42-ca503a6737b4" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/d83acd41-67d5-440f-9f2b-4b7d871ed5ab" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/f03fb796-73e2-4eac-a3d3-64c3e843e745" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/a6e13c8b-7b58-43c1-9ee2-b6b29ab04187" />

Some of these problems might just be because it is an unconditional model, which is why we are already returning focus to conditional models #202 . I think I'll just close this issue, and if Issue #202 shows promise, we can worry about increasing model capacity in that issue.

**Comment by fwilson12** [comment]
Although this issue is already closed, here are the results from the increased-capacity unconditional model I was training on 20k 16x16 MMLV samples:

<img width="1185" height="303" alt="Image" src="https://github.com/user-attachments/assets/d6e670ee-e078-44d7-ab98-f69ca54be584" />

##
Pretty much the same homogeneity and incoherence that we saw with the 32x32 samples

**Comment by schrum2** [comment]
Thanks for the update. I think Issue #202 shows that we really need the text conditioning to steer the model towards more coherence, and recent code changes will hopefully fix the speed issue I complained about above. This issue will remain closed, but the path foward is clear: lots of samples, with some kind of captioning ... possibly with higher model capacity, though I think the issue is more in the training signal than the model weights.
